### PR TITLE
Various fixes and refactoring of huffman compression/decompression

### PR DESCRIPTION
**Pull request by Robyt3**
See commit messages.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [X] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or Valgrind's memcheck](https://github.com/ddnet/ddnet/blob/master/docs/DEBUGGING.md#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)
- [ ] I didn't use generative AI to generate more than single-line completions:
   - AI found the unsigned integer underflow on decompression, but assumed it could cause very many loop iterations (DoS). However, in practice, tree node lookup prevents the loop from being iterated often. I fuzzed the decompression for a combined ~20 hours with AFL++ and could not find any crashes or hangs with ASAN, UBSAN, CFISAN and CmpLog.
   - AI found a potential OOB write in the compression function that can only happen when the output size is zero, which is now checked with assertions.
   - AI noted the missing handling of negative sizes.

### CODE REVIEW

File: src/engine/shared/huffman.cpp [hunk 1]
```diff
@@ -193,8 +193,12 @@ int CHuffman::Compress(const void *pInput, int InputSize, void *pOutput, int Out
 	HUFFMAN_MACRO_LOADSYMBOL(HUFFMAN_EOF_SYMBOL);
 	HUFFMAN_MACRO_WRITE();
 
-	// write out the last bits
-	*pDst++ = Bits;
+	// write out the last bits if we have any
+	if(Bitcount != 0)
+	{
+		dbg_assert(pDst < pDstEnd, "No space for last bits. Bitcount=%d Bits=%d", Bitcount, Bits);
```
Comment (heinrich5991) [inline src/engine/shared/huffman.cpp]: Should return an error instead, like above. Right now, this creates a DoS vulnerability, funnily. ^^

File: src/test/huffman_test.cpp [hunk 2]
```diff
@@ -71,7 +71,7 @@ TEST(Huffman, CompressionTrailingNull)
 
 	const int CompressedSize = Huffman.Compress(aInput, sizeof(aInput), aCompressed, sizeof(aCompressed));
 
-	const unsigned char aExpected[] = {0xBE, 0xFD, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x15, 0x37, 0x00};
+	const unsigned char aExpected[] = {0xBE, 0xFD, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x15, 0x37};
```
Comment (heinrich5991) [inline src/test/huffman_test.cpp:93]: Yea, I also stumbled upon this edge case when writing the huffman code in `libtw2`.

File: src/engine/shared/huffman.cpp [hunk 3]
```diff
@@ -253,6 +253,10 @@ int CHuffman::Decompress(const void *pInput, int InputSize, void *pOutput, int O
 		{
 			// remove the bits that the lut checked up for us
 			Bits >>= HUFFMAN_LUTBITS;
+			if(Bitcount < HUFFMAN_LUTBITS)
+			{
+				return -1;
+			}
```
Comment (heinrich5991) [inline src/engine/shared/huffman.cpp:264]: For correctness, this implicitly relies on the fact that the EOF symbol is longer than `HUFFMAN_LUTBITS`. I think the code shouldn't make that assumption.

File: src/engine/shared/huffman.cpp [hunk 4]
```diff
@@ -93,7 +93,7 @@ void CHuffman::ConstructTree(const unsigned *pFrequencies)
 	Setbits_r(m_pStartNode, 0, 0);
 }
 
-void CHuffman::Init(const unsigned *pFrequencies)
+void CHuffman::Init()
```
Comment (heinrich5991) [inline src/engine/shared/huffman.cpp:94]: Sounds sensible.

**Review by heinrich5991 — COMMENTED** [review]
Sensible changes, but we need to be careful here, not introducing bugs when fixing hypothetical ones. ^^

File: src/engine/shared/huffman.cpp (hunk 1 above)
Comment (Robyt3) [reply src/engine/shared/huffman.cpp]: For more defensive programming in this case or do you know a case where this can happen? I couldn't find any combination of input and output for which this condition is not satisfied (except with the already disallowed `OutputSize == 0`) so I made it an assertion instead of returning an error.

File: src/engine/shared/huffman.cpp (hunk 3 above)
Comment (Robyt3) [reply src/engine/shared/huffman.cpp:264]: Do you have a suggestion to fix this?

I also noticed we aren't checking for underflow on `Bitcount -= pNode->m_NumBits;` in the branch above yet.

File: src/engine/shared/huffman.cpp (hunk 1 above)
Comment (heinrich5991) [reply src/engine/shared/huffman.cpp]: I assumed this can happen. I wouldn't want to add this for defensive programming. Why do you think this cannot happen?

This "just" needs a compression where the last output bits go into one byte past the end of the buffer. Am I overlooking something?

Comment (Robyt3) [reply src/engine/shared/huffman.cpp]: Bytes are only written to `pDst` when the `HUFFMAN_MACRO_WRITE` macro is used, and in this location to handle the last remaining bits.

https://github.com/ddnet/ddnet/blob/e7ebc4fdb419f2f48aaff41d8f2daf99f3cba948/src/engine/shared/huffman.cpp#L144-L155

The `HUFFMAN_MACRO_WRITE` macro checks whether `pDstEnd` was reached only *after* writing the current byte. So when we reach this `dbg_assert`, the previous use of `HUFFMAN_MACRO_LOADSYMBOL(HUFFMAN_EOF_SYMBOL); HUFFMAN_MACRO_WRITE();` should ensure that `pDst < pDstEnd` as long as `OutputSize >= 1`. However, it seems the correctness of this also implicitly depends on the encoding of `HUFFMAN_EOF_SYMBOL` taking 8 or more bits, otherwise the `while(Bitcount >= 8)` is never entered so `pDst == pDstEnd` would also be unchecked.

Comment (Robyt3) [reply src/engine/shared/huffman.cpp]: Added `if(pDst == pDstEnd) return -1;` here and changed the `HUFFMAN_MACRO_WRITE` macro to check before writing the byte.

Also added a test to check for correct compression of input data of size `0`.

**Comment by jxsl13** [comment]
btw the last entry of the dictionary/frequencies is never used afair `1517`

**Comment by Robyt3** [comment]
> btw the last entry of the dictionary/frequencies is never used afair `1517`

Thanks, fixed.

File: src/engine/shared/huffman.cpp (hunk 3 above)
Comment (Robyt3) [reply src/engine/shared/huffman.cpp:264]: > For correctness, this implicitly relies on the fact that the EOF symbol is longer than `HUFFMAN_LUTBITS`.

Please elaborate. If the EOF symbol is not longer than `HUFFMAN_LUTBITS`, then the initial lookup should yield the EOF node immediately and we enter the `if(pNode->m_NumBits)` branch instead of this else-branch.

Comment (heinrich5991) [reply src/engine/shared/huffman.cpp:264]: You're right.

**Review by heinrich5991 — APPROVED** [review]
Thanks. The separate commits made it a lot easier to review the changes. :)
